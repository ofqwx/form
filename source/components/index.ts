export { FormContext } from './Form/FormProvider';
export { default as FormProvider } from './Form/FormProvider';
export { default as Field } from './Field/Field';
export { default as Form } from './Form/Form';
