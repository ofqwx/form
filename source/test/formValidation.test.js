/**
 * @jest-environment jsdom
 */

import React from 'react';
import { Form, Input } from '../components';
import { fireEvent, render, screen } from '@testing-library/react';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import '@testing-library/jest-dom/extend-expect';
import validations from '../validation/validation';

expect.extend({ toMatchDiffSnapshot });

test('String required validation', async () => {
  const { asFragment } = render(
    <Form initialValues={{ firstName: 'antonio' }} onSubmit={() => undefined}>
      <Input
        label="First name"
        name="firstName"
        validations={[validations.string.required('First name is required')]}
      />

      <button type="submit">Submit</button>
    </Form>
  );

  const firstRender = asFragment();

  fireEvent.change(screen.getByRole('textbox'), {
    target: { value: '' },
  });

  // Display error when value is empty after change
  expect(firstRender).toMatchDiffSnapshot(asFragment());

  fireEvent.change(screen.getByRole('textbox'), {
    target: { value: 'Wooga' },
  });

  // Remove error when value is not empty after change
  expect(firstRender).toMatchDiffSnapshot(asFragment());
});

test('String regex validation', async () => {
  const regex = /^(?:wooga\.name).[\w|\W]{1,}/;

  const { asFragment } = render(
    <Form onSubmit={() => undefined}>
      <Input
        label="First name"
        name="firstName"
        validations={[
          validations.string.regex(
            'First name must start with "wooga.name"',
            regex
          ),
        ]}
      />

      <button type="submit">Submit</button>
    </Form>
  );

  const firstRender = asFragment();

  fireEvent.change(screen.getByRole('textbox'), {
    target: { value: 'antonio' },
  });

  // Display error if value doesn't match with the regex
  expect(firstRender).toMatchDiffSnapshot(asFragment());

  fireEvent.change(screen.getByRole('textbox'), {
    target: { value: 'wooga.name antonio' },
  });

  // Remove error if value match with the regex
  expect(firstRender).toMatchDiffSnapshot(asFragment());
});

test('String regex and required validation', async () => {
  const regex = /^(?:wooga\.name).[\w|\W]{1,}/;

  const { asFragment } = render(
    <Form initialValues={{ firstName: 'antonio' }} onSubmit={() => undefined}>
      <Input
        label="First name"
        name="firstName"
        validations={[
          validations.string.required('First name is required'),
          validations.string.regex(
            'First name must start with "wooga.name"',
            regex
          ),
        ]}
      />

      <button type="submit">Submit</button>
    </Form>
  );

  const firstRender = asFragment();

  fireEvent.change(screen.getByRole('textbox'), {
    target: { value: '' },
  });

  // Display error if value is empty
  expect(firstRender).toMatchDiffSnapshot(asFragment());

  fireEvent.change(screen.getByRole('textbox'), {
    target: { value: 'wooga' },
  });

  // Remove required error and display regex error
  expect(firstRender).toMatchDiffSnapshot(asFragment());

  fireEvent.change(screen.getByRole('textbox'), {
    target: { value: 'wooga.name.antonio' },
  });

  // Remove all errors, value is not empty and match regex
  expect(firstRender).toMatchDiffSnapshot(asFragment());
});
