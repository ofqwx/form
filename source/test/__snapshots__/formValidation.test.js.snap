// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`String regex and required validation 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -6,12 +6,18 @@
        </label>
        <br />
        <input
          name=\\"firstName\\"
          type=\\"text\\"
-         value=\\"antonio\\"
+         value=\\"\\"
        />
+       <br />
+       <div
+         style=\\"color: red;\\"
+       >
+         First name is required
+       </div>
      </fieldset>
      <button
        type=\\"submit\\"
      >
        Submit"
`;

exports[`String regex and required validation 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -6,12 +6,18 @@
        </label>
        <br />
        <input
          name=\\"firstName\\"
          type=\\"text\\"
-         value=\\"antonio\\"
+         value=\\"wooga\\"
        />
+       <br />
+       <div
+         style=\\"color: red;\\"
+       >
+         First name must start with \\"wooga.name\\"
+       </div>
      </fieldset>
      <button
        type=\\"submit\\"
      >
        Submit"
`;

exports[`String regex and required validation 3`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -6,11 +6,11 @@
        </label>
        <br />
        <input
          name=\\"firstName\\"
          type=\\"text\\"
-         value=\\"antonio\\"
+         value=\\"wooga.name.antonio\\"
        />
      </fieldset>
      <button
        type=\\"submit\\"
      >"
`;

exports[`String regex validation 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -6,12 +6,18 @@
        </label>
        <br />
        <input
          name=\\"firstName\\"
          type=\\"text\\"
-         value=\\"\\"
+         value=\\"antonio\\"
        />
+       <br />
+       <div
+         style=\\"color: red;\\"
+       >
+         First name must start with \\"wooga.name\\"
+       </div>
      </fieldset>
      <button
        type=\\"submit\\"
      >
        Submit"
`;

exports[`String regex validation 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -6,11 +6,11 @@
        </label>
        <br />
        <input
          name=\\"firstName\\"
          type=\\"text\\"
-         value=\\"\\"
+         value=\\"wooga.name antonio\\"
        />
      </fieldset>
      <button
        type=\\"submit\\"
      >"
`;

exports[`String required validation 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -6,12 +6,18 @@
        </label>
        <br />
        <input
          name=\\"firstName\\"
          type=\\"text\\"
-         value=\\"antonio\\"
+         value=\\"\\"
        />
+       <br />
+       <div
+         style=\\"color: red;\\"
+       >
+         First name is required
+       </div>
      </fieldset>
      <button
        type=\\"submit\\"
      >
        Submit"
`;

exports[`String required validation 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -6,11 +6,11 @@
        </label>
        <br />
        <input
          name=\\"firstName\\"
          type=\\"text\\"
-         value=\\"antonio\\"
+         value=\\"Wooga\\"
        />
      </fieldset>
      <button
        type=\\"submit\\"
      >"
`;
